content/api/servo-assembly/arduino-commands.md

```diff
@@ -0,0 +1,196 @@
+// With the exception of the name (here it's `echo`), the rest of the function's
+// "signature" must be exactly as it is on the following line:
+static CommandError echo(int commandId, String argument) {
+
+  if (argument.length() == 0) {
+    // If something went wrong, you should return a `COMMAND_ERROR` with a
+    // message about the failure. That message will be fed back to Python where
+    // an exception will be raised.
+    return COMMAND_ERROR("nothing to echo");
+  }
+
+  while (argument.length() > 0) {
+    // You can use `pop_option` to extract individual parameters from the
+    // `argument` string. They're sent space-separated, so be sure not to
+    // include any spaces within your argument values!
+    //
+    // You don't need to check the length of `argument` before calling
+    // `pop_option` -- if there isn't an option to be had you'll get back an
+    // empty string ("").
+    String next_arg = pop_option(argument);
+
+    // You can write values to be read by the code on the Pi using the
+    // `serialWrite` function. The first two arguments need to be exactly as
+    // they are here; the third argument needs to be a `String` containing the
+    // value you want to send back to the Pi.
+    serialWrite(commandId, '>', next_arg);
+  }
+
+  // If everything went fine, you should return `OK`.
+  return OK;
+}
+```
+<!-- TODO: mention being able to send back comments? -->
+
+For other examples of how to define commands (and the helper functions you can
+use), you can have a look at the other commands the Arduino already supports.
+In particular the `get_version`, `led` and `servo`  are a good starting point.
+
+<!-- TODO should we error in robot-api/robotd if the incoming arguments include spaces? -->
+
+### Registering the command
+
+To register this command we need to modify the list of commands stored in the
+`commands` array. To do this we add a new line containing a `CommandHandler` for
+the command.
+
+```cpp
+static const CommandHandler commands[] = {
+  // ... lots of existing commands ... (make sure you leave them in place!)
+  CommandHandler("echo", &echo, "echoes its arguments, each on a new line"),
+};
+```
+
+The first argument is a quoted `String` literal containing the name you'll use
+in Python to call this command. It can contain letters, numbers and dashes (`-`).
+
+The second argument is the function you defined for your function, with an
+ampersand (`&`) just before the name.
+
+The third argument is another quoted `String` literal containing a description
+of the command. This helps document what the command is for, and will be
+returned as part of the message if you call the `help` command.
+
+## Calling commands
+
+To call a command from Python you should call the `direct_command` method on a
+`Servo` instance. You can send strings and numbers:
+
+```python
+r.servo_board.direct_command('echo', "first argument", 2, "third argument")
```

Comment:
Any reason for the mixing of quote types here?

Reply:
`'echo'` is for the machine and is close to being an enum value or a slug. The others are less so, though are arguably still much more that than my usual use of double quotes (which is for text intended for humans).

They should probably all be `'`.

Reply:
d1f728d

Reply:
Yeah, I'd agree than in typical applications the arguments to the command won't be intended to be human-readable text - more likely to be numbers or keywords (e.g. 'low' / 'high').